rfcs/0xx-tei_adapter/README.md

```diff
@@ -0,0 +1,93 @@
+to the TEI files in a TEI store, and not the full XML. This has the advantage that we don't have to manage an 
+ever growing S3 bucket (we don't have a way of cleaning up old versions of a record) but it has some disadvantages:
+- GitHub outages (which are rare but happen) might affect our reindexes.
+- We might be throttled during a reindex (?)
+- We go to the public internet for every TEI file for every reindex as opposed to getting each file exactly once 
+  per every change that affects it.
+  
+For these reasons, the proposal here is to store thw whole TEI file in a full VHS.
+### Store structure
+
+Therefore, the proposed store structure would be:
+
+| Id(bnumber) | Payload (location in s3) | deleted (true/false) | version |
+| ----------- | ------------------------ | ---------------------| --------|
+
+### Updates from github
+
+There are two possible way to receive regular updates from GitHub that I can think of:
+
+#### Use GitHub push notifications:
+  
+This means setting up a Webhook integration in the TEI repository that points to a URL of an API Gateway endpoint. 
+API Gateway calls a Lambda that forwards the JSON form GitHub to the TEI adapter. The advantages of this approach are: 
+- Usage of Webhooks in GitHub is very established 
+- it's very simple to implement. 
+  
+
+The biggest disadvantage is that we don't have control or visibility if the Webhook call to our API Gateway endpoint
+  fails or if there is an outage on GitHub. That may cause updates to not be propagated through in the pipeline
+  without us knowing.
+
+#### Pull the changes from GitHub at regular intervals:
+  
+This means having a Lambda that regularly requests changes from GitHub withing subsequent time intervals. 
+This is an approach we already use in the Sierra adapter and in the Calm adapter.
+The big advantage of this approach is that we control what happens if one request fail, how we should be notified
+and what action we should take.
+
+The proposal in this RFC is to use a pull approach so that we can be confident that all updates have either gone through
+successfully, or we are notified of any failure, and we can retry.
+
+This is a diagram of the proposed architecture: ![](architecture.png)
+
+The Lambda in the picture should generate windows and send them to the TEI adapter
+
+### TEI Adapter
+![](tei_adapter_stack.png)
+The proposal is to make the TEI adapter 2 services
+#### GitHub Api service
+
+It receives time windows from the lambda and for each one:
+  
+- It calls the GitHub API to get the commits within that window
+- Call the GitHub API to get the files changed in each commit
+- Send a message for each file
+
+
+####  TEI Adapter service
+
+It receives file URLs from the GitHub Api service and for each one:
+
+- Retrieves it from GitHub & store it in the TEI store
+- Send each record stored to the transformer
+
+### Full reharvest
+
+For the full reharvest we crawl the repo and send the URL of each file to the TEI Adaper service.
```

Comment:
What does "crawl the repo" mean in this context?

I wonder if looking at Git trees might be useful? Quick Git recap: you have "blobs" (files) and "trees". Trees form a recursive data structure:

<img width="547" alt="image" src="https://user-images.githubusercontent.com/301220/115731581-27de8280-a37f-11eb-94ca-15e85169aa20.png">

You can see the root tree here: https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/trees/master

If you used the trees to build a file map (path -> sha), that's another way to do it. Have some code that compares the current tree to a stored tree in VHS or whatever, and use that to identify files that need updating.

(Noticing deleted files gets a bit trickier in this case.)

So your process would go something like:

```
[get current tree] -> [compare tree to stored tree] -> [get each different file]
```

with the advantage that doing a full reharvest is just "get the very latest tree", and this runs in constant-time, rather than slowing down if there are more commits.

OTOH, you now have to hold an entire file map somewhere, hmm.

Reply:
This is where a bit more detail on the GitHub API calls would be useful, I wonder a bit about something like https://stackoverflow.com/questions/26925312/github-api-how-to-compare-2-commits

Reply:
Thanks, this is super helpful!

Reply:
Clearly, I have spent a lot less time elaborating the reharvest cause it's not that intersting usually. It's a thing that runs once or a handful of times and we don't usually make it fully production quality (they're run locally, they have no tests, no dedicated infra etc). But it is a bit laconic so I will expand on that. I was thinking to use the tree (I didn't know you could get that from the github api, thanks for that), but basically go through each file and send each one to the TEI adapter service, by passing the GIthub service entirely. This bit is a bit crucial to understand what was going on in my head, I realise, and I didn't think to add it 🤦‍♀️